src/main/scala/com/fulcrumgenomics/bam/Bams.scala

```diff
@@ -263,6 +298,54 @@ object Bams extends LazyLogging {
     new SelfClosingIterator(iter, () => queryIterator.close())
   }
 
+  /** Return an iterator over records sorted and grouped into [[Template]] objects. Although a queryname sort is
+    * guaranteed, the sort order may not be consistent with other queryname sorting implementations, especially in other
+    * tool kits. See [[com.fulcrumgenomics.bam.Bams.templateIterator]] for a [[Template]] iterator which emits templates
+    * in a non-guaranteed sort order.
+    *
+    * @see [[com.fulcrumgenomics.bam.Bams.templateIterator]]
+    *
+    * @param in a SamReader from which to consume records
+    * @param maxInMemory the maximum number of records to keep and sort in memory, if sorting is needed
+    * @param tmpDir a temp directory to use for temporary sorting files if sorting is needed
+    * @return an Iterator of queryname sorted Template objects
+    */
+  def sortedTemplateIterator(in: SamSource,
+                             maxInMemory: Int = MaxInMemory,
+                             tmpDir: DirPath = Io.tmpDir): SelfClosingIterator[Template] = {
+    sortedTemplateIterator(in.iterator, in.header, maxInMemory, tmpDir)
+  }
+
+  /** Return an iterator over records sorted and grouped into [[Template]] objects. Although a queryname sort is
+    * guaranteed, the sort order may not be consistent with other queryname sorting implementations, especially in other
+    * tool kits. See [[com.fulcrumgenomics.bam.Bams.templateIterator]] for a [[Template]] iterator which emits templates
+    * in a non-guaranteed sort order.
+    *
+    * @see [[com.fulcrumgenomics.bam.Bams.templateIterator]]
+    *
+    * @param iterator an iterator from which to consume records
+    * @param header the header associated with the records
+    * @param maxInMemory the maximum number of records to keep and sort in memory, if sorting is needed
+    * @param tmpDir a temp directory to use for temporary sorting files if sorting is needed
+    * @return an Iterator of queryname sorted Template objects
+    */
+  def sortedTemplateIterator(iterator: Iterator[SamRecord],
+                             header: SAMFileHeader,
+                             maxInMemory: Int,
+                             tmpDir: DirPath): SelfClosingIterator[Template] = {
+    val queryIterator = querySortedIterator(iterator, header, maxInMemory, tmpDir)
+
+    val _iterator = new Iterator[Template] {
+      override def hasNext: Boolean = queryIterator.hasNext
+      override def next: Template   = {
+        require(hasNext, "next() called on empty iterator")
+        Template(queryIterator)
+      }
+    }
+
+    new SelfClosingIterator(_iterator, () => queryIterator.close())
+  }
```

Comment:
So this method is here so that we can have an iterator over `Template`s but guaranteed to be in query-name order?  Can you explain a use-case?